src/Umbraco.Core/Services/ContentService.cs

```diff
@@ -382,13 +382,15 @@ public IEnumerable<IContent> GetByIds(IEnumerable<int> ids)
         /// <returns><see cref="IContent"/></returns>
         public IContent GetById(Guid key)
         {
-            using (var uow = UowProvider.GetUnitOfWork(readOnly: true))
-            {
-                var repository = RepositoryFactory.CreateContentRepository(uow);
-                var query = Query<IContent>.Builder.Where(x => x.Key == key);
-                var contents = repository.GetByQuery(query);
-                return contents.SingleOrDefault();
-            }
+            // the repository implements a cache policy on int identifiers, not guids,
+            // and we are not changing it now, but we still would like to rely on caching
+            // instead of running a full query against the database, so relying on the
+            // id-key map, which is fast.
+            //
+            // we should inject the id-key map but ... breaking changes ... yada

```

Comment:
We should (and can) inject this - we are allowed to change services and constructors. This should not break any site, it may break unit tests but that is ok. Relying on ApplicationContext.Current inside a service is super ugly